File: lab04.1/ciphers_python/trisemus_cipher.py
CYRILLIC_ALPHABET = "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ"
LATIN_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"


def _get_table_chars(*, keyword: str, alphabet: str) -> list[str]:
    table_chars = []
    for ch in (keyword + alphabet):
        if ch not in table_chars:
            table_chars.append(ch)
    return table_chars


def _make_table(*, keyword: str, alphabet: str, table_size: int | None = None) -> list[list[str]]:
    table_chars = _get_table_chars(keyword=keyword, alphabet=alphabet)
    if table_size is None:
        table_size = len(set(keyword))
    table = [[]]
    for ch in table_chars:
        if len(table[-1]) >= table_size:
            table.append([])
        table[-1].append(ch)
    return table


def _find_table_index(ch: str, table: list[list[str]]) -> int | tuple[int, int]:
    for i in range(len(table)):
        for j in range(len(table[i])):
            if table[i][j] == ch:
                return (i, j)
    return -1


def _trisemus_cipher(*, old_message: str, keyword: str, alphabet: str, step: int, table_size: int | None = None) -> str:
    table_upper = _make_table(
        keyword=keyword, 
        alphabet=alphabet.upper(), 
        table_size=table_size
    )
    table_lower = _make_table(
        keyword=keyword, 
        alphabet=alphabet.lower(), 
        table_size=table_size
    )
    step = int(step / abs(step))
    new_message = ""
    for old_ch in old_message:
        current_table = table_upper if old_ch.isupper() else table_lower
        old_ch_index = _find_table_index(old_ch, current_table)
        if old_ch_index == -1:
            new_ch = old_ch
        else:
            new_ch_index_i = old_ch_index[0] + step
            new_ch_index_j = old_ch_index[1]
            if new_ch_index_i == -1:
                new_ch_index_i = len(current_table) - 1
            if new_ch_index_i >= len(current_table) or new_ch_index_j >= len(current_table[new_ch_index_i]):
                new_ch_index_i = 0
            new_ch = current_table[new_ch_index_i][new_ch_index_j]
        new_message += new_ch
    return new_message


def trisemus_cipher_encryption(*, decrypted_message: str, keyword: str, alphabet: str, table_size: int | None = None) -> str:
    return _trisemus_cipher(
        old_message=decrypted_message, 
        keyword=keyword, 
        alphabet=alphabet,
        step=1,
        table_size=table_size,
    )


def trisemus_cipher_decryption(*, encrypted_message: str, keyword: str, alphabet: str, table_size: int | None = None) -> str:
    return _trisemus_cipher(
        old_message=encrypted_message, 
        keyword=keyword, 
        alphabet=alphabet,
        step=-1,
        table_size=table_size,
    )


def trisemus_cipher_cyrillic_encryption(decrypted_message: str, keyword: str, table_size: int | None = None) -> str:
    return trisemus_cipher_encryption(
        decrypted_message=decrypted_message,
        keyword=keyword,
        alphabet=CYRILLIC_ALPHABET,
        table_size=table_size,
    )


def trisemus_cipher_cyrillic_decryption(encrypted_message: str, keyword: str, table_size: int | None = None) -> str:
    return trisemus_cipher_decryption(
        encrypted_message=encrypted_message,
        keyword=keyword,
        alphabet=CYRILLIC_ALPHABET,
        table_size=table_size,
    )


def trisemus_cipher_latin_encryption(decrypted_message: str, keyword: str, table_size: int | None = None) -> str:
    return trisemus_cipher_encryption(
        decrypted_message=decrypted_message,
        keyword=keyword,
        alphabet=LATIN_ALPHABET,
        table_size=table_size,
    )


def trisemus_cipher_latin_decryption(encrypted_message: str, keyword: str, table_size: int | None = None) -> str:
    return trisemus_cipher_decryption(
        encrypted_message=encrypted_message,
        keyword=keyword,
        alphabet=LATIN_ALPHABET,
        table_size=table_size,
    )

File: lab04.1/ciphers_csharp/TrisemusCipher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ciphers_csharp {
    internal static class TrisemusCipher {
        private const string CYRILLIC_ALPHABET = "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ";
        private const string LATIN_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

        private static List<char> _getTableChars(string keyword, string alphabet) {
            List<char> tableChars = new List<char>();
            foreach (char ch in (keyword + alphabet)) {
                if (!tableChars.Contains(ch)) {
                    tableChars.Add(ch);
                }
            }
            return tableChars;
        }

        private static List<List<char>> _makeTable(string keyword, string alphabet, int? tableSize = null) {
            List<char> tableChars = _getTableChars(keyword, alphabet);
            if (tableSize == null) {
                tableSize = keyword.Distinct().ToList().Count;
            }
            List<List<char>> table = new List<List<char>>() { new List<char>() };
            foreach (char ch in tableChars) {
                if (table.Last().Count >= tableSize) {
                    table.Add(new List<char>());
                }
                table.Last().Add(ch);
            }
            return table;
        }

        private static (int, int) _findTableIndex(char ch, List<List<char>> table) {
            for (int i = 0; i < table.Count; i++) { 
                for (int j = 0;  j < table[i].Count; j++) {
                    if (table[i][j] == ch) {
                        return (i, j);
                    }
                }
            }
            return (-1, -1);
        }

        private static string _trisemusCipher(string oldMessage, string keyword, string alphabet, int step, int? tableSize = null) {
            List<List<char>> tableUpper = _makeTable(keyword, alphabet.ToUpper(), tableSize);
            List<List<char>> tableLower = _makeTable(keyword, alphabet.ToLower(), tableSize);
            step = (int)(step / Math.Abs(step));
            string newMessage = "";
            foreach (char oldCh in oldMessage) {
                List<List<char>> currentTable = char.IsUpper(oldCh) ? tableUpper : tableLower;
                (int, int) oldChIndex = _findTableIndex(oldCh, currentTable);
                char newCh;
                if (oldChIndex == (-1, -1)) {
                    newCh = oldCh;
                }
                else {
                    int newChIndexI = oldChIndex.Item1 + step;
                    int newChIndexJ = oldChIndex.Item2;
                    if (newChIndexI == -1) {
                        newChIndexI = currentTable.Count - 1;
                    }
                    if (newChIndexI >= currentTable.Count || newChIndexJ >= currentTable[newChIndexI].Count) {
                        newChIndexI = 0;
                    }
                    newCh = currentTable[newChIndexI][newChIndexJ];
                }
                newMessage += newCh;
            }
            return newMessage;
        }

        public static string TrisemusCipherEncryption(string decryptedMessage, string keyword, string alphabet, int? tableSize = null) {
            return _trisemusCipher(decryptedMessage, keyword, alphabet, 1, tableSize);
        }

        public static string TrisemusCipherDecryption(string encryptedMessage, string keyword, string alphabet, int? tableSize = null) {
            return _trisemusCipher(encryptedMessage, keyword, alphabet, -1, tableSize);
        }

        public static string TrisemusCipherCyrillicEncryption(string decryptedMessage, string keyword, int? tableSize = null) {
            return TrisemusCipherEncryption(decryptedMessage, keyword, CYRILLIC_ALPHABET, tableSize);
        }

        public static string TrisemusCipherCyrillicDecryption(string encryptedMessage, string keyword, int? tableSize = null) {
            return TrisemusCipherDecryption(encryptedMessage, keyword, CYRILLIC_ALPHABET, tableSize);
        }

        public static string TrisemusCipherLatinEncryption(string decryptedMessage, string keyword, int? tableSize = null) {
            return TrisemusCipherEncryption(decryptedMessage, keyword, LATIN_ALPHABET, tableSize);
        }

        public static string TrisemusCipherLatinDecryption(string encryptedMessage, string keyword, int? tableSize = null) {
            return TrisemusCipherDecryption(encryptedMessage, keyword, LATIN_ALPHABET, tableSize);
        }
    }
}
